presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -146,6 +192,228 @@ public boolean hasOrderBy()
+
+        private DistinctingAccumulator(
+                Accumulator accumulator,
+                List<Type> inputTypes,
+                List<Integer> inputs,
+                Optional<Integer> maskChannel,
+                Session session,
+                JoinCompiler joinCompiler)
+        {
+            this.accumulator = requireNonNull(accumulator, "accumulator is null");
+            this.maskChannel = requireNonNull(maskChannel, "maskChannel is null");
+
+            hash = new MarkDistinctHash(session, inputTypes, Ints.toArray(inputs), Optional.empty(), joinCompiler, UpdateMemory.NOOP);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return hash.getEstimatedSize() + accumulator.getEstimatedSize();
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return accumulator.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void addInput(Page page)
+        {
+            // 1. filter out positions based on mask, if present
+            Page filtered = maskChannel
+                    .map(channel -> filter(page, page.getBlock(channel)))
+                    .orElse(page);
+
+            // 2. compute a distinct mask
+            Work<Block> work = hash.markDistinctRows(filtered);
+            checkState(work.process());
+            Block distinctMask = work.getResult();
+
+            // 3. feed a Page with a new mask to the underlying aggregation
+            accumulator.addInput(filtered.prependColumn(distinctMask));
+        }
+
+        @Override
+        public void addInput(WindowIndex index, List<Integer> channels, int startPosition, int endPosition)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void addIntermediate(Block block)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void evaluateIntermediate(BlockBuilder blockBuilder)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void evaluateFinal(BlockBuilder blockBuilder)
+        {
+            accumulator.evaluateFinal(blockBuilder);
+        }
+    }
+
+    private static Page filter(Page page, Block mask)
+    {
+        int[] ids = new int[mask.getPositionCount()];
+        int next = 0;
+        for (int i = 0; i < page.getPositionCount(); ++i) {
+            if (BOOLEAN.getBoolean(mask, i)) {
+                ids[next++] = i;
+            }
+        }
+
+        return page.getPositions(ids, 0, next);
+    }
+
+    private static class DistinctingGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final GroupedAccumulator accumulator;
+        private final MarkDistinctHash hash;
```

Comment:
Which columns does it use to mark rows? Does it really save any memory in comparison to `MarkDistinctNode` just below FA

Reply:
It's not about saving memory, but avoiding shuffles. See my comment above.

Although, thinking about it, it might save some global memory, too. It uses the columns in the DISTINCT arguments + a synthetic column computed by the aggregation operator containing a group number (BIGINT). MarkDistinct uses the DISTINCT arguments plus all the columns used in the group-by clause.